lib/Driver.cpp

```diff
@@ -315,3 +318,84 @@ Driver::Driver(Diagnostics &diagnostics, const Configuration &config, const Proc
+  if (configuration.coverageInfo.empty()) {
+    return nullptr;
+  }
+  llvm::Expected<std::unique_ptr<llvm::coverage::CoverageMapping>> maybeMapping =
+      llvm::coverage::CoverageMapping::load({ configuration.executable },
+                                            configuration.coverageInfo);
+  if (!maybeMapping) {
+    std::string error;
+    llvm::raw_string_ostream os(error);
+    llvm::logAllUnhandledErrors(maybeMapping.takeError(), os, "Cannot read coverage info: ");
+    diagnostics.warning(os.str());
+    return nullptr;
+  }
+  return std::move(maybeMapping.get());
+}
+
+std::vector<FunctionUnderTest> Driver::getFunctionsUnderTest(std::vector<Test> &tests) {
+  std::vector<FunctionUnderTest> functionsUnderTest;
+
+  singleTask.execute("Gathering functions under test", [&]() {
+    std::unique_ptr<llvm::coverage::CoverageMapping> coverage = loadCoverage(config, diagnostics);
+    if (coverage) {
+      /// Some of the function records contain just name, the others are prefixed with the filename
+      /// to avoid collisions
+      /// TODO: check case when filename:functionName is not enough to resolve collisions
+      /// TODO: pick a proper data structure
+      std::unordered_map<std::string, std::unordered_set<std::string>> scopedFunctions;
+      std::unordered_set<std::string> unscopedFunctions;
+      for (auto &it : coverage->getCoveredFunctions()) {
+        if (!it.ExecutionCount) {
+          continue;
+        }
+        std::string scope;
+        std::string name = it.Name;
+        size_t idx = name.find(':');
+        if (idx != std::string::npos) {
+          scope = name.substr(0, idx);
+          name = name.substr(idx + 1);
+        }
+        if (scope.empty()) {
+          unscopedFunctions.insert(name);
+        } else {
+          scopedFunctions[scope].insert(name);
+        }
+      }
+      for (auto &test : tests) {
+        for (auto &bitcode : program.bitcode()) {
+          for (llvm::Function &function : *bitcode->getModule()) {
+            bool covered = false;
+            std::string name = function.getName().str();
+            if (unscopedFunctions.count(name)) {
+              covered = true;
+            } else {
+              std::string filepath = SourceLocation::locationFromFunction(&function).unitFilePath;
+              std::string scope = llvm::sys::path::filename(filepath).str();
+              if (scopedFunctions[scope].count(name)) {
+                covered = true;
+              }
+            }
+            if (covered) {
+              functionsUnderTest.emplace_back(&function, &test, 1);
+            }
+          }
+        }
+      }
+    } else {
+      for (auto &test : tests) {
+        for (auto &bitcode : program.bitcode()) {
+          for (llvm::Function &function : *bitcode->getModule()) {
+            functionsUnderTest.emplace_back(&function, &test, 1);
+          }
+        }
+      }
+    }
+  });
+
+  return functionsUnderTest;
+}
```

Comment:
This should obviously not be here, but I'll extract and move it away as soon as the dead code is gone.